### PR TITLE
feat(ai): configure custom provider HTTP headers securely

### LINKED ISSUE #3135: [Feature] Support custom HTTP headers for Custom AI Providers
## Problem / pain point
Some OpenAI-compatible and self-hosted AI gateways require additional HTTP headers beyond the configured endpoint, API key, and model (for example, `HTTP-Referer`, `X-Title`, tenant/project identifiers, or provider-specific routing headers). Netcatty's Custom AI Provider setup currently cannot send these headers, so those providers cannot be used without modifying a proxy.

## Proposed solution
Add an optional **Custom HTTP Headers** section when creating or editing a Custom AI Provider. Let users add multiple key/value header pairs, apply them to every request for that provider, and preserve them when the provider is edited. Header values that are secrets should be stored with the same protection used for API keys.

## Alternatives considered
Using a local reverse proxy to inject headers works, but adds infrastructure, configuration, and another failure point. Choosing only providers that do not require extra headers is not practical.

## Related area
AI assistant

## How important is this to you?
Would improve my daily workflow

## Before submitting
- [x] I searched existing issues and discussions for similar requests

### PR DESCRIPTION
**Pull request by binaricat**
## Summary

Custom AI providers can now add, edit, and remove HTTP headers in settings. The same headers reach chat, model discovery, and connection testing. Values are masked in the editor and protected using the existing credential bridge before saving; unavailable encryption blocks saving.

## Type of Change

- [x] New feature

## Related Issue (optional)

Closes #3135

## Changes Made

- Reuse the existing `customHeaders` request support and resolve encrypted values at the main-process request boundary, including case-insensitive authentication overrides.
- Validate header names, values, and duplicate names; reject transport-owned Content-Length and Transfer-Encoding. Preserve saved headers and model metadata when reopening the editor.
- Carry headers through the existing encrypted cloud-sync payload, re-encrypt on import, and preserve local headers for older settings-only payloads. Explicit empty headers propagate removals.
- Keep the existing API-key requirement and advanced-parameter controls unchanged.

## Screenshots / Demo

Mounted the actual provider settings form in an isolated Vite page with a test credential/fetch bridge and exercised it in Ego browser: add a header, connection test, save, reopen, edit, and reject duplicate names. The browser intercepted the final remove-button click; the mounted React regression test separately clicks Remove and Save and verifies the empty saved headers. Confirmed masked values, encrypted saved output, decrypted outgoing headers, and nonzero input layout. Browser screenshot capture timed out; no screenshot is attached. A local HTTP server independently verified decrypted request headers from the actual main-process injection function.

## Testing

- [x] Local UI exercised through Vite (isolated form with test bridges).
- [x] ESLint passes for changed source and tests.
- [x] 174 focused AI, provider-form, synchronization, backup, migration, and reasoning tests pass, plus the real HTTP regression.
- [x] Mounted React regression test verifies async header loading blocks Save, preserves model metadata, and supports removing/saving headers.
- [x] Main-process HTTP regression verifies decrypted headers, case-insensitive auth override, and failure when credentials cannot be unlocked.
- [x] Final-head GitHub checks all succeeded: lint-and-test, macOS, Windows, Linux x64, and Linux ARM64 packages.
- [ ] Repository typecheck: compared changed files against original main-branch content with the same compiler configuration: exactly the same 7 pre-existing diagnostics before and after; no new diagnostics. Full repository also has unrelated existing errors.
- [x] No capability catalog changes.

Two independent reviewers report CLEAN on `210879b8482e4a2f367ea6b3effff11dcf281a69`. Provider header encryption now finishes in the existing protected preparation phase, before any vault import, settings write, or interrupted-apply marker. All automatic/manual sync, restore, and migration paths reuse its prepared result. Failure tests verify no partial imports/settings changes or lingering recovery marker; success tests verify one encryption pass and unchanged input. Migration initialization and first-publication failures are also covered. The existing recovery protection mechanism is unchanged. Latest remote Codex review reports no major issues; all review threads are resolved. Merged after the coordinating task verified all final-head checks.

Changed-file lint and the isolated local build pass. Main integration preserves provider reasoning settings, tab-bar position, and right-click long-press behavior and tests.

## Checklist

- [x] Follows existing project style and credential/request mechanisms.
- [x] In-form help documents hidden/encrypted values and secure-storage errors.
- [x] Existing provider configurations remain supported.

Integration: current main is incorporated. Header configuration, provider reasoning settings, tab-bar position, and right-click long-press synchronization and their tests are all preserved.

### CODE REVIEW

**Comment by binaricat** [comment]
@codex review

**Comment by binaricat** [comment]
@codex review

File: infrastructure/ai/providerHeaderCredentials.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    return [name, encrypted];
+  })));
+}
+
+export async function decryptProviderHeaders(headers: Record<string, string> = {}): Promise<Record<string, string>> {
+  return Object.fromEntries(await Promise.all(Object.entries(headers).map(async ([name, value]) => {
+    const result = await decryptFieldResult(value);
+    if (result.unread) throw new Error('Unable to decrypt custom HTTP headers');
+    return [name, result.value ?? ''];
```
[the remark this answers is not in the document]
Comment (binaricat) [reply infrastructure/ai/providerHeaderCredentials.ts:40]: Reviewed this against the pre-change tree (`e7d7a12a9`) and with both independent local reviewers. Before this PR, non-test `customHeaders` references existed only in the type declaration and request-building helper; there was no settings create/edit writer. We found no supported prior UI flow that persisted plaintext headers. Such existing data would require manual storage edits or a handcrafted imported payload.

This PR encrypts every new header save and every incoming header-bearing sync/restore payload, including converting legacy plaintext on save/import. Accepting plaintext reads preserves existing manually configured providers; it does not introduce a plaintext write path. Loading-time migration would additionally require coordinating asynchronous writes in both provider-state hooks and is pre-existing hardening outside this narrowly scoped feature, absent evidence of a prior supported writer. Keeping it out of #3135 rather than adding speculative startup mutation. If a concrete legacy writer or affected profile is identified, that migration can be assessed separately.

**Comment by binaricat** [comment]
@codex review

File: components/settings/tabs/ai/ModelSelector.tsx [hunk 2]
```diff
@@ -95,12 +95,13 @@ export const ModelSelector: React.FC<{
   presetModels?: readonly string[];
   placeholder?: string;
   apiKey?: string;
+  customHeaders?: Record<string, string>;
```
[the remark this answers is not in the document]
Comment (binaricat) [reply components/settings/tabs/ai/ModelSelector.tsx:98]: The requested #3135 feature is explicitly additional headers “beyond the configured endpoint, API key, and model”; it does not ask to replace API-key authentication. This PR deliberately retains the pre-existing API-key requirement consistently in the editor, probe, catalog, and chat.

I re-ran actual SDK generation with a custom Authorization header: no API key produces AI_LoadAPIKeyError before any bridge request; with the configured key the request reaches the bridge and the custom header overrides auth as tested. Both independent reviewers rechecked this against the issue and confirmed it is not a regression introduced by this PR. Broadening only the gate would be incorrect; adding header-only authentication for protocol-specific or arbitrary vendor headers is a separate feature. No change to the established authentication contract in #3135.

File: application/state/useAISettingsState.ts [hunk 3]
```diff
@@ -302,6 +303,27 @@ export function useAISettingsState() {
...
+    let cancelled = false;
+    migrateLegacyProviderHeaders(
+      localStorageAdapter.read<ProviderConfig[]>(STORAGE_KEY_AI_PROVIDERS) ?? [],
+    )
+      .then((migrated) => {
+        if (cancelled || migrated == null) return;
+        // Write storage first so cross-window sync readers never observe the
+        // migration result without its encryption.
+        localStorageAdapter.write(STORAGE_KEY_AI_PROVIDERS, migrated);
+        setProvidersRaw(migrated);
```
[the remark this answers is not in the document]
Comment (binaricat) [reply application/state/useAISettingsState.ts]: Confirmed: the bot-added migration in dd5749460 could overwrite newer provider settings. Rather than extend startup state coordination for a migration already rejected as outside #3135, reverted the entire bot-added commit in e10e5ccbc. The resulting product tree is identical to the independently reviewed and tested 688f54403. Neither state hook performs the migration anymore, so the stale-snapshot write path is removed. New header saves and sync/import paths retain encryption protection.

**Comment by binaricat** [comment]
@codex review

File: infrastructure/ai/providerHeaderCredentials.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+/** Validate before requests as well as saving. HTTP names are case insensitive. */
+export function parseProviderHeaderRows(rows: HeaderRow[]): Record<string, string> {
+  const entries: [string, string][] = [];
+  const names = new Set<string>();
+  for (const row of rows) {
+    if (!row.name && !row.value) continue;
+    const name = row.name.trim();
+    if (!/^[!#$%&'*+.^_`|~0-9A-Za-z-]+$/.test(name)
+      || /[^\t\x20-\x7e\x80-\xff]/.test(row.value)
+      || names.has(name.toLowerCase())) {
```
[the remark this answers is not in the document]
Comment (binaricat) [reply infrastructure/ai/providerHeaderCredentials.ts:15]: Confirmed with an actual Node HTTP server: combining Transfer-Encoding with the automatically generated Content-Length is rejected with HPE_INVALID_CONTENT_LENGTH / HTTP 400. Fixed the editor validator to reject both Content-Length and Transfer-Encoding case-insensitively before any save/probe, with regression inputs and localized explanation that these headers are managed automatically. No transport or authentication redesign. Focused validator and mounted-form tests pass; independently re-reviewed.

**Comment by binaricat** [comment]
@codex review

**Comment by binaricat** [comment]
@codex review

File: application/syncPayload.ts [hunk 5]
```diff
@@ -825,7 +835,12 @@ async function applySyncableSettings(settings: NonNullable<SyncPayload['settings
   const ai = settings.ai;
   if (ai) {
     if (ai.providers != null) {
-      const providers = await Promise.all(ai.providers.map(withLocalEncryptedApiKey));
+      const providers = await Promise.all(ai.providers.map(async (provider) => {
+        const next = await withLocalEncryptedApiKey(provider);
+        return isRecord(provider.customHeaders)
+          ? { ...next, customHeaders: await encryptProviderHeaders(await decryptProviderHeaders(provider.customHeaders as Record<string, string>)) }
```
[the remark this answers is not in the document]
Comment (binaricat) [reply application/syncPayload.ts]: Fixed in db0238ce6. Header credentials are now prepared immediately after payload sanitization, before any vault import or settings write, and the prepared encrypted values are reused when saving. Added regression coverage for sync and restore encryption failure (no imports or settings changes), plus successful single encryption without mutating the source payload. Targeted tests, HTTP coverage, lint, build, and two independent reviews pass. This change is limited to preventing partial data application; it does not redesign the outer recovery sentinel lifecycle.

**Comment by binaricat** [comment]
@codex review

File: application/syncPayload.ts [hunk 6]
```diff
@@ -1158,6 +1170,15 @@ function applyPayload(
   // so a previously poisoned cloud/backup snapshot can still restore host
   // shells and let the user re-enter secrets (#2702).
   const sanitizedPayload = stripSyncPayloadEncryptedCredentials(payload);
+  // Prepare credentials before any vault or settings mutation. Encryption can
+  // fail when local credential storage is unavailable; reuse the result below.
+  const providers = sanitizedPayload.settings?.ai?.providers;
+  const preparedProviders = providers == null ? undefined : await Promise.all(providers.map(async (provider) => {
+    const next = await withLocalEncryptedApiKey(provider);
+    return isRecord(provider.customHeaders)
+      ? { ...next, customHeaders: await encryptProviderHeaders(await decryptProviderHeaders(provider.customHeaders as Record<string, string>)) }
```
[the remark this answers is not in the document]
Comment (binaricat) [reply application/syncPayload.ts:1179]: Fixed in 210879b84. Header preparation now returns an apply callback consumed by the existing protected prepareApply phase. Both automatic-sync paths, both settings-sync paths, local restore, and both migration initialization/publication phases prepare credentials before the sentinel is written, then reuse the prepared result. The protected wrapper and sentinel lifecycle are unchanged. Real protected-apply tests cover header encryption failure and success for sync/restore, including no lingering sentinel, single encryption, and unchanged source payload. Migration tests also exercise failure during both initial apply and first publication. 174 targeted tests, lint, build, and two independent full-call-chain reviews pass.

**Comment by binaricat** [comment]
@codex review
